[Docs] Clarify caveats for phonetic filters replace option

cbuescher opened the pull request:
The `replace` option in the phonetic token filter can have suprising side
effects, e.g. such as described in #26921. This PR adds a note to be mindful
about such scenarios and offers alternatives to using the `replace` option.

Closes #26921

Commit:
[Docs] Clarify caveats for phonetic filters replace option

The `replace` option in the phonetic token filter can have suprising side
effects, e.g. such as described in #26921. This PR adds a note to be mindful
about such scenarios and offers alternatives to using the `replace` option.

Closes #26921

elasticmachine commented:
Pinging @elastic/es-search

jtibshirani reviewed (commented):
Thanks @cbuescher! I made note of a typo and also had a small suggestion: it might be helpful to add one line to the last paragraph of the [match query docs about fuzziness](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html#query-dsl-match-query-fuzziness), noting that the issue applies not only to synonyms but to any token filter that produces multiple tokens at the same position.

Commit:
Fixing typos

Commit:
Rephrase

cbuescher commented:
@jtibshirani thanks for the review, I corrected the typo, added some links and reprhased the paragraph a bit. 

I don't completely get your comment about adding to the last paragraph of the match query docs about fuzziness. The issue this documentation is trying to adress is to be aware of the fact that stacked token like the synonym filter produces can have unwanted side-effects, in this case most notably that match queries skip applying fuzziness to synonym queries the the original query is rewritten to when we have stacked token in the same position. The paragraph you linked talks about generating phrases which sounds related, but in the case of the example that is given in https://github.com/elastic/elasticsearch/issues/26921#issuecomment-335087296 the `auto_generate_synonyms_phrase_query` doesn't really make a difference in the output. I have problems relating the two things, so did you have any specific in minf to add to the  "match query docs about fuzziness" paragraph?

jtibshirani commented on docs/plugins/analysis-phonetic.asciidoc:
```diff
@@ -65,6 +65,14 @@ GET phonetic_sample/_analyze
+It is important to note that `"replace": false` can lead to unexpected behavior since
+the original and the phonetically analyzed version are both kept at the same token position.
+Some queries handle these stacked tokens in special ways. For example, the fuzzy `match`
+query does not apply <<fuzziness>> to stacked synonym tokens. This can lead to issues that are
```
I tried to build these plugin docs, but I get the error " IDREF attribute linkend references an unknown ID "fuzziness"".

cbuescher replied:
Yep, you are right, doesn't work here as well. I changed it to using "{ref}" which is what I think we use in cases where we link between two "books" in the docs.

jtibshirani reviewed (commented):
Sorry for the confusion, I meant the last paragraph in the section on fuzziness (not the last paragraph of the document overall). The paragraph ends with "Note that fuzzy matching is not applied to terms with synonyms, as under the hood these terms are expanded to a special synonym query that blends term frequencies, which does not support fuzzy expansion.".

This is just a suggestion, happy to approve + merge after the build issue is fixed.

cbuescher commented:
>  I meant the last paragraph in the section on fuzziness

Thanks for the clarification, that makes a lot of sense to me now. Will look into it tomorrow, also into the plugin-docs issue you see (although the tests seem to pass, so I guess we're good)

jtibshirani commented:
👍 will take another look tomorrow. I'm not sure our docs check actually runs all docs builds. For example, in the past I accidentally introduced an issue into the HLRC docs although the PR passed the docs checks.

Commit:
Merge branch 'master' into fix-26921

Commit:
Add note to match-query docs

Commit:
Adding cross-book link

cbuescher commented:
@jtibshirani added the note and corrected the link, this should be good to go now once the tests pass.
